src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -0,0 +1,248 @@
+                durationCancellationSource.CancelAfter(ruleDuration.Value);
+            }
+
+            try
+            {
+                bool completePipeline = false;
+                while (!completePipeline)
+                {
+                    TaskCompletionSource<object> triggerSatisfiedSource =
+                        new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                    ICollectionRuleTrigger trigger = null;
+                    try
+                    {
+                        trigger = factory.Create(
+                            _endpointInfo,
+                            () => triggerSatisfiedSource.TrySetResult(null),
+                            _ruleOptions.Trigger.Settings);
+
+                        if (null == trigger)
+                        {
+                            throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_TriggerFactoryFailed, _ruleOptions.Trigger.Type));
+                        }
+
+                        // Start the trigger.
+                        await trigger.StartAsync(linkedToken).ConfigureAwait(false);
+
+                        // The pipeline signals that it has started just after starting a non-startup trigger.
+                        // Instances with startup triggers signal start after having finished executing the action list.
+                        if (trigger is not ICollectionRuleStartupTrigger)
+                        {
+                            // Signal that the pipeline trigger is initialized.
+                            _startedSource.TrySetResult(null);
+                        }
+
+                        // Wait for the trigger to be satisfied.
+                        await triggerSatisfiedSource.WithCancellation(linkedToken).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        try
+                        {
+                            // Intentially not using the linkedToken. If the linkedToken was signaled
+                            // due to pipeline duration expiring, try to stop the trigger gracefully
+                            // unless forced by a caller to the pipeline.
+                            await trigger.StopAsync(token).ConfigureAwait(false);
+                        }
+                        finally
+                        {
+                            if (trigger is IAsyncDisposable asyncDisposableTrigger)
+                            {
+                                await asyncDisposableTrigger.DisposeAsync().ConfigureAwait(false);
+                            }
+                            else if (trigger is IDisposable disposableTrigger)
+                            {
+                                disposableTrigger.Dispose();
+                            }
+                        }
+                    }
+
+                    DateTime currentTimestamp = DateTime.Now;
```

Comment:
UtcNow. This can actually do strange things like go back in time on daylight savings time etc.